### ISSUE TITLE
[no-unnecessary-condition] Ignore loops with generator yields.

### ISSUE THREAD

**Issue by brainkim** [typescript-eslint/typescript-eslint#1067]
**Repro**
```JSON
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "sourceType": "module",
    "createDefaultProgram": true
  },
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```ts
function *ints() {
  let i = 0;
  while (true) {
    yield i++;
  }
}
```

**Expected Result**
No error.
**Actual Result**
```
Unnecessary conditional, value is always truthy  @typescript-eslint/no-unnecessary-condition
```
**Additional Info**
Given that `no-unnecessary-condition` is supposed to be a stronger version of `no-constant-condition`, it would be nice if it followed eslint’s choice to treat constant-condition loops with generator yields as valid code.
Related eslint PRs:
- https://github.com/eslint/eslint/pull/8762
- https://github.com/eslint/eslint/pull/10827
Related issues:
- https://github.com/typescript-eslint/typescript-eslint/issues/1015
- https://github.com/typescript-eslint/typescript-eslint/issues/1016
**Versions**
package | version
-- | --
@typescript-eslint/eslint-plugin | 2.3.3
@typescript-eslint/parser | 2.3.3
TypeScript | 3.6.4
ESLint | 6.5.1
node | 10.16.3
npm | 6.9.0

**Comment by brainkim** [comment]
@bradzacher I’m not sure if this a plugin rule option. ESLint simply disables the rule if the loop contains a yield operator by default.

**Comment by bradzacher** [comment]
TBF eslint's version is also less strict. I can understand some people might see having `while (true) { yield ... }` as a bug, as an infinite loop is a subset of the problem space for loop conditions.

I.e. you could have a bug like:
```
for (const i of ints()) {
  // infinite loop due to infinite yield
}
```

Having it behind an option gives someone the choice of allowing this pattern in their codebase, as opposed to forcing it on everyone.

**Comment by brainkim** [comment]
I’m okay with this behavior being behind an option, but would recommend that this be the default behavior insofar as 1. this matches analogous eslint behavior and 2. `while (true) { yield ... }` is idiomatic, regularly used code in generators/async generators.

I also don’t think “infinite yields” should be considered a bug, more like a consequence of the halting problem.
